### DISCUSSION TITLE
Surprising behaviour of extension method resolution

### DISCUSSION THREAD

**Discussion by bernd5** [dotnet/csharplang#7678] [Q&A]
Compile the following [code](https://sharplab.io/#v2:EYLgHgbALANALiATgVwHYB8ACAmAjAWAChMAGAAk1ygG4jSyBlAewFsBTOACwEtUBzWoSIAxJkwB0AIQCGicQEFxAYWkAbVQFEwcALIdOTACYAKVMnUBKQUVTT2AZwAO0gMZsyoiTMREA3kTJAimwyeQCg/0Ig6IpcCDJeODZEW1UyYDE0lXUtXX0jY2dEO3tYkgBtAF0y+wtwmLJIhobMAHYa8QBJewA5c1UAeUQNFkc4AE9jK3qYgF8Z+cJFmzs2J1d3Tz8ZzABmWPicMiUmdTYXOG4mVHtcthurm5mm5oB6V4B3bi4mZDgyH72dxsbT3eyPMjsH6GUrMKE8fgKRDFcZ3B7XexdXr9IYjMbjAA8ABUAHwwMgfX6qQzpdy8ABuTAA1mxDOS4GJxDNonsDulMmRun11LjRhNiSTjFxuKVOiczhdHhKAPxkFynVTnS7XOpRZovZpBNpqjVaiEAXnNZDM6jI6HQJoV2tQyl+qH+lrIJEEhuWepa+0o8Qyp0F2JFwzFhNJUp4soASmxpIYBqhVON5ZrFdcVY6s87dYaDYaKO11U6LVabWl7XmzddXWgPVbvdygotoos/bYHM43IxWPpeHxtv6KIG4sFQsjpKjQeinmPizyJ8GBUKcZH8RLYzKyESqqrZCjCyXAuaSWRj7OyJ7q3aHdfxuIADL3PhcW8tn3toizIA===):

```csharp
using System;
using System.Collections.Generic;
using Something;

Foo.Bar.A.CallExtMethod(null);

namespace Foo.Bar
{
    //using Something; //<-- with that it would work
    class A
    {
        static internal bool CallExtMethod(params string[] strings)
        {
            return strings.IsNullOrEmpty();
        }
    }
}

namespace Foo
{
    public static class CollectionsExtensions
    {
        //without those extension methods Something.ArrayExtensions.IsNullOrEmpty<T>, would be invoked, too.
        public static bool IsNullOrEmpty<T>(this ICollection<T>? collection)
        {
            return collection == null || collection.Count == 0;
        }

        public static bool IsNullOrEmpty<T>(this IReadOnlyCollection<T>? collection)
        {
            return collection == null || collection.Count == 0;
        }
    }
}

namespace Something
{
    public static class ArrayExtensions
    {
        public static bool IsNullOrEmpty<T>(this T[]? array)
            => array == null || array.Length == 0;
    }
}
```

I expected that `Something.ArrayExtensions.IsNullOrEmpty<T>` would be invoked.
But the method search algorithm seems to stop searching for extension methods if multiple candidates where found in current or parent namespace. It seems to continue than with namespaces in using clauses.

Is that behaviour intended?

**Comment by DavidArno** [comment]
Yes, that behaviour is intended. Extension method resolution works on "closeness" and will stop as soon as it finds a suitable match in the closest possible location. There's a [great article by Eric Lippert called Closer is better](https://ericlippert.com/2013/12/23/closer-is-better/) that explains the mechanism nicely.
